### PR TITLE
feat: add support for addtl endpoints

### PR DESCRIPTION
**Pull request by wass3r**
closes https://github.com/go-vela/community/issues/188

1. add /authenticate endpoint support to exchange code/state for tokens
2. add /authenticate/token support to auth via PAT
3. add timeout on default http client
4. tests

### CODE REVIEW

File: vela/authentication.go [hunk 1]
```diff
@@ -94,3 +96,93 @@ func (svc *AuthenticationService) RefreshAccessToken(refreshToken string) (*Resp
...
+	// set the received access token
+	svc.token = &at
+
+	return at, resp, err
+}
+
+// ExchangeTokens handles the last part of the OAuth flow. It uses the supplied
+// code and state values to attempt to exchange them for Vela Access and
+// Refresh tokens.
+func (svc *AuthenticationService) ExchangeTokens(opt *OAuthExchangeOptions) (string, string, *Response, error) {
```
[the remark this answers is not in the document]
Comment (wass3r) [reply vela/authentication.go:139]: fixed

**Review by KellyMerrick — DISMISSED** [review]
lgtm

**Review by matt-fevold — COMMENTED** [review]
How do we normally handle breaking changes?

**Comment by wass3r** [comment]
maybe it was wrong for me to call out, but the outcome of the change is that users will have to update their token. arguably, the server changes are breaking as well in that case. 

but it's a good but kind of a loaded question. we don't have a precedent, i think. but we should probably define what that _should_ look like in the future. lots of aspects in terms of communication, deprecation periods, etc

**Comment by wass3r** [comment]
in fact, in reviewing the changes again, the callout is probably misplaced here. (it was more applicable with an earlier set of change). the previous functionality is still there and basically works like it did before. you pass and set the token the same way. the blocker will be at the server which will return with invalid token.

**Comment by kneal** [comment]
The change looks good to me!

As far as the deprecation goes, I've seen us do this in other areas of the code:
https://github.com/go-vela/types/blob/4495c6cec9e4b0e9c97af87e069acb86af31a99c/yaml/step.go#L92-L96

I think that's a nice way of putting it and makes it large and in your face so we don't forget. 

My personal, thoughts are I think we should remove the dead code this time vs provide a warning. I like the warning in most cases but honestly, auth is kind of a big deal. I don't think most of the changes in this release are backward compatible so if someone is running an older version of the server it would be safer for them to stick with the SDK version that's certified for that server version.

**Comment by wass3r** [comment]
Yea, I agree with that. I pushed a new commit to remove the now obsolete function. It resulted in having to update a few other things 🐙

**Review by kneal — APPROVED** [review]
LGTM 🐬

**Review by jbrockopp — APPROVED** [review]
LGTM

**Pull request merged**